src/farcasterd/syncer_state_machine.rs

```diff
@@ -0,0 +1,327 @@
+                blockchain,
+                network,
+                &runtime.config,
+            )? {
+                event.complete_ctl_service(service_id, Request::SyncerTask(syncer_task))?;
+                Ok(Some(SyncerStateMachine::AwaitingSyncerRequest(
+                    AwaitingSyncerRequest {
+                        source,
+                        syncer_task_id,
+                        syncer: ServiceId::Syncer(blockchain, network),
+                    },
+                )))
+            } else {
+                Ok(Some(SyncerStateMachine::AwaitingSyncer(AwaitingSyncer {
+                    source,
+                    syncer: ServiceId::Syncer(blockchain, network),
+                    syncer_task: syncer_task,
+                    syncer_task_id,
+                })))
+            }
+        }
+
+        _ => Ok(None),
+    }
+}
+
+fn transition_to_awaiting_syncer_request(
+    event: Event,
+    _runtime: &mut Runtime,
+    awaiting_syncer: AwaitingSyncer,
+) -> Result<Option<SyncerStateMachine>, Error> {
+    let AwaitingSyncer {
+        source,
+        syncer,
+        syncer_task,
+        syncer_task_id,
+    } = awaiting_syncer;
+    match (event.request.clone(), event.source.clone()) {
+        (Request::Hello, syncer_id) if syncer == syncer_id => {
+            event.complete_ctl_service(syncer.clone(), Request::SyncerTask(syncer_task))?;
+            Ok(Some(SyncerStateMachine::AwaitingSyncerRequest(
+                AwaitingSyncerRequest {
+                    source,
+                    syncer_task_id,
+                    syncer,
+                },
+            )))
+        }
+        (req, source) => {
+            if let Request::Hello = req {
+                trace!(
+                    "Request {} from {} invalid for state awaiting syncer.",
+                    req,
+                    source
+                );
+            } else {
+                warn!(
+                    "Request {} from {} invalid for state awaiting syncer.",
+                    req, source
+                );
+            }
+            Ok(Some(SyncerStateMachine::AwaitingSyncer(AwaitingSyncer {
+                source,
+                syncer,
+                syncer_task,
+                syncer_task_id,
+            })))
+        }
+    }
+}
+
+fn transition_to_end(
+    mut event: Event,
+    runtime: &mut Runtime,
+    awaiting_syncer_request: AwaitingSyncerRequest,
+) -> Result<Option<SyncerStateMachine>, Error> {
+    let AwaitingSyncerRequest {
+        syncer_task_id,
+        source,
+        syncer,
+    } = awaiting_syncer_request;
+    match (event.request.clone(), event.source.clone()) {
+        (Request::SyncerEvent(SyncerEvent::SweepSuccess(success)), syncer_id)
+            if syncer == syncer_id && success.id == syncer_task_id =>
+        {
+            if let Some(Some(txid)) = success
+                .txids
+                .clone()
+                .pop()
```

Comment:
don't quite understand why this is matching on the last tx only, or receiving a vec of txs at all? don't see why it's not a single tx.

Reply:
This goes back to the syncer handling a potential excess of transaction outputs for Monero, since Monero transactions can handle only a limited number of inputs. We use the same type to report Success in the bitcoin syncer, so we can just pop the first element, since we don't handle multiple sweep transactions for Bitcoin.

Reply:
ok, so it's again there to keep the implementations symmetric - fair